frontend/catalyst/utils/jax_extras.py

```diff
@@ -477,3 +576,372 @@ def make_jaxpr_f(*args, **kwargs):
+            ]
+            aval = aval.update(shape=tuple(shape))
+        out_tracers.append(maker(aval))
+    return out_tracers
+
+
+TracerLike = TypeVar("TracerLike")
+
+
+def infer_input_type_unshared(inputs: List[TracerLike]) -> InputType:
+    """Infer the input type of a function having `inputs` Jax tracers."""
+
+    def _is_tracer_like(x):
+        return hasattr(x, "aval")
+
+    expl_ins = inputs
+    impl_avals = []
+    expl_avals = []
+    for o in expl_ins:
+        assert _is_tracer_like(o), (o,)
+        if isinstance(o.aval, DShapedArray):
+            shape2 = []
+            for d in o.aval.shape:
+                if _is_tracer_like(d):
+                    shape2.append(DBIdx(len(impl_avals)))
+                    impl_avals.append(d.aval)
+                else:
+                    shape2.append(d)
+            expl_avals.append(o.aval.update(shape=tuple(shape2)))
+        else:
+            expl_avals.append(o.aval)
+    return (*[(i, False) for i in impl_avals], *[(i, True) for i in expl_avals])
+
+
+@dataclass
+class ExpansionStrategy:
+    """Describe the settings affecting the Jax dyanmic API tracing of the nested programs.
+    Args:
+        axes_specs: Axes specification used to convert static dimensions to the dynamic ones
+        input_unshare_variables: Treat each dynamic dimension as a distinct dimension, even if some
+                                 of them are described by the same dimension variables.
+        output_include_indbidx_vars: Include variables corresponding to the InDBIdx references in
+                                     the result
+        output_force_arg0_outdbidx: Force all references pointing to the first arguments to be
+                                    OutDBIdx. This is typically required to avoid input references
+                                    to the loop iteration variable.
+    """
+
+    axes_specs: Sequence[AbstractedAxesSpec] | None
+    input_unshare_variables: bool
+    output_include_indbidx_vars: bool
+    output_force_arg0_outdbidx: bool
+
+
+def while_loop_expansion_strategy(preserve_dimensions=False):
+    """Arguments and results expansion strategy for while-loops."""
+    return ExpansionStrategy(None, not preserve_dimensions, True, False)
+
+
+def for_loop_expansion_strategy(preserve_dimensions=False):
+    """Arguments and results expansion strategy for for-loops."""
+    return ExpansionStrategy(None, not preserve_dimensions, True, True)
+
+
+def cond_expansion_strategy():
+    """Arguments and results expansion strategy for conditionals."""
+    return ExpansionStrategy(None, False, True, False)
+
+
+def infer_output_type(
+    constants: List[TracerLike],
+    expanded_inputs: List[TracerLike],
+    outputs: List[TracerLike],
+    expansion_strategy: ExpansionStrategy,
+    num_implicit_inputs: int | None = None,
+) -> Tuple[List[TracerLike], OutputType]:
+    """Deduce the Jax ``out_type`` given input and ouputs abstract entities. By abstract entities
```

Comment:
Because JAX code is quite difficult to read, I think it would be very helpful if a docstring like this would not just say that it is deducing the output type, but also what that means concretely (for example "the output type should be this in so-and-so situation" or "the output type is deduced from the input types by doing so-and-so", or whatever the idea the programmer had in mind when writing this function).

Reply:
The output type is an explicitly defined core entity of Jax, so I am assuming that the reader knows what it is. See the definition [here](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L1310). So, I don't think that this function is the best place for the descriptions like this. Should we add a separate documentation/glossary on Jax into our dev docs and add a link to it here?

Reply:
My comment is not asking to define what an output type is, but to explain what this function does. The short description is "it deduces output types from input types", but that is not enough imo for a maintainer to understand what is going on (which is super important!).
When you wrote this function, you had an understanding of what the function should be doing, which you then translated into code. What I'm asking for is a small number of sentences that could help someone reading the code quickly grasp this "understanding" you had of "what the function should be doing".

Reply:
I see, please check the new comment.

Reply:
Thanks, the added comments are super helpful :)